### PR TITLE
Go example plugin

### PR DESCRIPTION
**Pull request by qlikmats**
Fixes #[issue number].

### Status
```
[x] Under development
[x] Waiting for code review
[ ] Waiting for merge
```
### Information
```
[ ] Contains breaking changes
[ ] Contains new API(s)
[x] Contains documentation
[ ] Contains test
[ ] Contains examples
```
### To-do list
```
[ ] 
[ ] 
[ ] ...
```

### CODE REVIEW

File: examples/go/basic_example/server.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
+/**
+ * Implements the qlik.sse.Connector service.
+ */
+
+//go:generate protoc -I ..\..\..\proto ..\..\..\proto\ServerSideExtension.proto --go_out=plugins=grpc:.\gen
```
Comment (carlioth) [inline examples/go/basic_example/server.go]: This is an issue when running on mac, change to //go:generate protoc -I ../../../proto ../../../proto/ServerSideExtension.proto --go_out=plugins=grpc:./gen
Guess that should work on Windows, please verify

File: examples/go/basic_example/server.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+		}
+	}
+}
+
+/*
+ * Main function.
+ */
+func main() {
+	flag.Parse()
+
```
Comment (carlioth) [inline examples/go/basic_example/server.go:256]: Add some initial information logging, what port it's running on.

**Review by carlioth — COMMENTED** [review]
Nice example. 
I've been trying it out to run containerized, some thoughts from me in comments. 
It would also be nice to add an qvf, consistent with the other examples.

File: examples/go/basic_example/server.go (hunk 1 above)
Comment (qlikmats) [reply examples/go/basic_example/server.go]: Indeed it works, and I have changed to forward-slash. Thanks.

File: examples/go/basic_example/server.go (hunk 2 above)
Comment (qlikmats) [reply examples/go/basic_example/server.go:256]: Good point. Will add that.

**Comment by carlioth** [comment]
Another thing to consider is to use dep, as dependency management. Also consider checking in you vendor (dependencies). You can read more / see example here: https://github.com/qlik-trial/cloud-app-retriever#dependency-management
More about dep here: https://hackernoon.com/using-go-dep-as-a-project-maintainer-641d1f3006d7

File: examples/go/README.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+This will generate the gRPC/Protobuf file(s) that your server will use  for implementing the qlik.sse.Connector service.
+
+
+
+### Configuring QlikSense to use the sample gRPC server
+By default, the Go sample plug-in runs on port 50051 on localhost, so for a QlikSense Desktop installation, the following should be added to settings.ini:
+
+[Settings 7] 
+
```
Comment (qlik-aoh) [inline examples/go/README.md]: Remove blank.

File: examples/go/README.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+Note that the string SSE_Go is the identifier that will prefix all plug-in functions when they are called from within Qlik.
+Use a different identifier for your own plug-in, and remember that this exact string has to be used for your Sense applications to work with the extension.
+
+The address (localhost:50051) should of course match the address in the server's configuration file.
+
+For single-machine development and testing it is OK to use unsecure communication, but for production scenarios you should use certificates. See [Generating certificates](../../generate_certs_guide/README.md).
+
+## Starting the server
+
+The Go gRPC sample server can be built and started using the go command:
```
Comment (qlik-aoh) [inline examples/go/README.md]: More of a script than a go command.
Also, we should aim for platform independence and the script is windows only.

File: examples/go/README.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+## Implementing a server - Protobuf generated files
+The interface between the Qlik Engine acting as a client and the Server-side extension acting as server is defined in 
+the file [ServerSideExtension.proto](../../proto/ServerSideExtension.proto). 
+
+
+## RPC methods
+The RPC methods implemented in the Basic example are GetCapabilities and ExecuteFunction. There is no script support.
+
+### GetCapabilities method
+The GetCapabilities method just returns a Capabilities object.
```
Comment (qlik-aoh) [inline examples/go/README.md]: That's not a very helpful description. Howabout

The GetCapabilities method returns a [Capabilities](https://github.com/qlik-oss/server-side-extension/blob/master/docs/SSE_Protocol.md#qlik.sse.Capabilities) object describing the operations supported by the plugin.

File: examples/go/basic_example/server.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"net"
+
+	"github.com/golang/protobuf/proto"
+
+	pb "github.com/qlikmats/server-side-extension/examples/go/basic_example/gen"
```
Comment (qlik-aoh) [inline examples/go/basic_example/server.go]: Shouldn't the path be relative rather than point out your fork repo?

**Review by qlik-aoh — CHANGES_REQUESTED** [review]
Perhaps you could some CircleCi test using this plugin?

File: examples/go/README.md (hunk 4 above)
Comment (qlikmats) [reply examples/go/README.md]: I was thinking of having Windows/Linux scripts, but I can scrap the scripts altogether and just document the command line(s).

File: examples/go/README.md (hunk 5 above)
Comment (qlikmats) [reply examples/go/README.md]: copy/paste "error" from one of the other examples, so I will change here and make a comment for the other plugin as well :)

File: examples/go/basic_example/server.go (hunk 6 above)
Comment (qlikmats) [reply examples/go/basic_example/server.go]: Don't think it should be relative, if I have understood Go import paths correctly, but it should not point to my fork when "merging" to the parent repo. For now, I think it needs to be like this in order to work...

**Review by qlik-aoh — APPROVED** [review]

File: examples/go/README.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+Next, you need to download and install the [protobuf compiler](https://github.com/google/protobuf/releases) and add the *protoc* executable to your %PATH%. Then, to install the protoc plugin for Go, execute:
+
+*go get -u github.com/golang/protobuf/protoc-gen-go*
+
+Note that the *protoc-gen-go* program also needs to be in your %PATH%.
+
+After setting up your [Go](https://golang.org/) environment, open a command prompt and execute: 
+
+*go get github.com/qlikmats/server-side-extension/examples/go/basic_example*
```
Comment (tobiaslindulf) [inline examples/go/README.md]: This link has to be changed before merging to qlik-oss.

File: examples/go/README.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+You will get an "error" saying there are no go files in the \gen folder, but that's  fine - we will generate a file in that folder in the following step.
+
+From your Go SSE plugin folder (i.e. %GOPATH%\src\github.com\qlikmats\server-side-extension\examples\go\basic_example), execute:
```
Comment (tobiaslindulf) [inline examples/go/README.md]: This link has to be changed before merging to qlik-oss.

File: examples/go/README.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
+# Writing an SSE plugin using Go
+
+General background information is found here:
+
+* [Writing an SSE Plugin](../../docs/writing_a_plugin.md)
+* [Protocol Description](../../docs/SSE_Protocol.md) (API reference)
+* Documentation and tutorials for [gRPC](http://www.grpc.io/docs/) and [protobuf](https://developers.google.com/protocol-buffers/docs/overview)
+
+## Getting started
```
Comment (tobiaslindulf) [inline examples/go/README.md]: Make the steps in here to a list of points instead. Easier to follow that way.

File: examples/go/README.md [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+First of all, you need an installation of the [Go](https://golang.org/) programming language. 
+
+Next, you need to download and install the [protobuf compiler](https://github.com/google/protobuf/releases) and add the *protoc* executable to your %PATH%. Then, to install the protoc plugin for Go, execute:
```
Comment (tobiaslindulf) [inline examples/go/README.md]: "Then" should be the start of a new point/step.

Comment (tobiaslindulf) [inline examples/go/README.md]: You can also link to the https://github.com/golang/protobuf#installation page where this is described in a good way.

File: examples/go/README.md [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+*go get -u github.com/golang/protobuf/protoc-gen-go*
+
+Note that the *protoc-gen-go* program also needs to be in your %PATH%.
```
Comment (tobiaslindulf) [inline examples/go/README.md]: I had problems to find the protoc-gen-go locally on windows after this command. Finally I discovered it ended up in C:\Users\\[User]\go\bin .

File: examples/go/README.md (hunk 10 above)
Comment (tobiaslindulf) [inline examples/go/README.md]: It was not obvious to me, which zip to choose from in the list. I had to scroll down quite long on that page to understand that I had to choose the windows zip.
Maybe state something like "download one of the pre-built binaries in zip packages: protoc-$VERSION-$PLATFORM.zip".

File: examples/go/README.md [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+From your Go SSE plugin folder (i.e. %GOPATH%\src\github.com\qlikmats\server-side-extension\examples\go\basic_example), execute:
+
+*go generate*
+
+This will generate the gRPC/Protobuf file(s) that your server will use  for implementing the qlik.sse.Connector service.
+
+
+
+### Configuring QlikSense to use the sample gRPC server
+By default, the Go sample plug-in runs on port 50051 on localhost, so for a QlikSense Desktop installation, the following should be added to settings.ini:
```
Comment (tobiaslindulf) [inline examples/go/README.md]: Without changing the code, port 50055 was opened, not 50051.

File: examples/go/README.md [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+*[Settings 7]* 
+*SSEPlugin=SSE_Go, localhost:50051[,PATH_TO_CERTIFICATE_FOLDER]*
```
Comment (tobiaslindulf) [inline examples/go/README.md]: Without changing the code, port 50055 was opened, not 50051.

Comment (tobiaslindulf) [inline examples/go/README.md]: The qvf app refers to SSEGo, not SSE_Go.

File: examples/go/basic_example/server.go [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+	AllowScript:      false,
+	PluginIdentifier: "SSE Go Plugin",
+	PluginVersion:    "1.0.0",
+	Functions:        functionDefinitions}
+
+/*
+ * Service impl.
+ */
+func (*server) GetCapabilities(context.Context, *pb.Empty) (*pb.Capabilities, error) {
+	for _, c := range capabilities.Functions {
```
Comment (tobiaslindulf) [inline examples/go/basic_example/server.go]: Add a logline saying log.Printf("GetCapabilities called")

File: examples/go/README.md [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+### GetCapabilities method
+The GetCapabilities method returns a Capabilities object, describing the operations supported by the plugin.
+
+### ExecuteFunction method
+The ExecuteFunction method switches over the numeric function identifier sent in the *qlik-functionrequestheader-bin* header. 
+Each case construct then iterates over the BundledRows elements packed into the request stream and writes the results to the output stream.  There are two functions implemented in the Go example plugin:
+
+ - *EchoString* (echoes the supplied string back to Sense)
+ - *SumOfRow* (summarizes two columns, in a row-wise manner)
+
```
Comment (tobiaslindulf) [inline examples/go/README.md]: To be consistent between our different examples I think we should add some more examples:
-SumOfCol: To have an aggregated example.
-Cache and NoCache functions to show how to manipulate headers.

**Review by tobiaslindulf — CHANGES_REQUESTED** [review]
Also add a Go section in https://github.com/qlikmats/server-side-extension/blob/master/README.md . Just like we have links to the other languages.
Consider updating the repo from qlik-oss master first to get the java changes in before merging (perhaps not needed).

File: examples/go/README.md (hunk 7 above)
Comment (qlikmats) [reply examples/go/README.md]: Yes, fully aware of that.

File: examples/go/README.md (hunk 8 above)
Comment (qlikmats) [reply examples/go/README.md]: Yes, fully aware of that.

File: README.md [hunk 16]
```diff
@@ -36,5 +36,10 @@ This repository includes documentation that describes the SSE protocol, how to u
 * [Writing an SSE Plugin using C#](examples/CSharp/README.md)
 * The [SSE R-plugin](https://github.com/qlik-oss/sse-r-plugin) is written in C#
 
+### Go
+* [A basic Go example](examples/go/basic_example/README.md)
+* [Writing an SSE Plugin using Go](examples/go/README.md)
```
Comment (qlikmats) [inline README.md]: Add this README.md

File: examples/go/README.md (hunk 9 above)
Comment (qlikmats) [reply examples/go/README.md]: Done.

File: examples/go/basic_example/server.go (hunk 14 above)
Comment (qlikmats) [reply examples/go/basic_example/server.go]: Done.

File: examples/go/basic_example/README.md [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+*[Settings 7]* 
+*SSEPlugin=SSEGo, localhost:50055[,PATH_TO_CERTIFICATE_FOLDER]*
+
+Note that the string SSEGo is the identifier that will prefix all plug-in functions when they are called from within Qlik.
+Use a different identifier for your own plug-in, and remember that this exact string has to be used for your Sense applications to work with the extension.
+
+The address (localhost:50055) should of course match the address in the server's configuration file.
+
+For single-machine development and testing it is OK to use unsecure communication, but for production scenarios you should use certificates. See [Generating certificates](../../generate_certs_guide/README.md).
```
Comment (josefinestal) [inline examples/go/basic_example/README.md]: This path does not seem to work. I think you might need another ../

File: examples/go/basic_example/README.md [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+The Go gRPC sample server can be built and started using the go command:
+
+*go run server.go*
+
+This will start the server in unsecure mode. To run in secure mode, execute the following command instead:
+
+*go run server.go -tls -cert_file=..\..\..\generate_certs_guide\sse_qliktest_generated_certs\sse_qliktest_server_certs\sse_server_cert.pem -key_file=..\..\..\generate_certs_guide\sse_qliktest_generated_certs\sse_qliktest_server_certs\sse_server_key.pem*
+
+## Run the example app
+Copy the file *SSE Go.qvf* to your Sense apps folder, i.e. *C:\Users\[user]\Qlik\Sense\Apps* and start Sense desktop (make sure the SSE Go plugin is up and running).
```
Comment (josefinestal) [inline examples/go/basic_example/README.md]: The dash before the "[" disappears when viewing the file. I think it's enough to add another "\" to solve it.

**Review by josefinestal — APPROVED** [review]
I mainly focused on the documentation and trying to run the example. A part from the comments I gave it looks good, and the example runs smoothly.

File: examples/go/README.md [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+## RPC methods
+The RPC methods implemented in the Basic example are GetCapabilities and ExecuteFunction. There is no script support.
+
+### GetCapabilities method
+The GetCapabilities method returns a Capabilities object, describing the operations supported by the plugin.
+
+### ExecuteFunction method
+The ExecuteFunction method switches over the numeric function identifier sent in the *qlik-functionrequestheader-bin* header. 
+Each case construct then iterates over the BundledRows elements packed into the request stream and writes the results to the output stream.  There are two functions implemented in the Go example plugin:
+
```
Comment (qlikmats) [inline examples/go/README.md]: "There are five functions..."

**Review by tobiaslindulf — APPROVED** [review]

**Pull request merged**